src/ir/possible-contents.cpp

```diff
@@ -525,21 +524,53 @@ struct InfoCollector
         return SignatureResultLocation{targetType, i};
       });
   }
-  void visitCallRef(CallRef* curr) {
-    auto targetType = curr->target->type;
+  template<typename T> void handleIndirectCall(T* curr, Type targetType) {
+    // If the type is unreachable, nothing can be called (and there is no heap
+    // type to get).
     if (targetType != Type::unreachable) {
-      auto heapType = targetType.getHeapType();
-      handleCall(
-        curr,
-        [&](Index i) {
-          return SignatureParamLocation{heapType, i};
-        },
-        [&](Index i) {
-          return SignatureResultLocation{heapType, i};
-        });
+      handleIndirectCall(curr, targetType.getHeapType());
     }
   }
 
+  void visitCall(Call* curr) {
+    Name targetName;
+    if (!Intrinsics(*getModule()).isCallWithoutEffects(curr)) {
+      // This is just a normal call.
+      handleDirectCall(curr, curr->target);
+      return;
+    }
+    // A call-without-effects receives a function reference and calls it, the
+    // same as a CallRef. When we have a flag for non-closed-world, we should
+    // handle this automatically by the reference flowing out to an import,
+    // which is what binaryen intrinsics look like. For now, to support use
+    // cases of a closed world but that also use this intrinsic, handle the
+    // intrinsic specifically here.
+    auto* target = curr->operands.back();
+    if (auto* refFunc = target->dynCast<RefFunc>()) {
+      // We can see exactly where this goes.
+      handleDirectCall(curr, refFunc->func);
+    } else {
+      // We can't see where this goes. We must be pessimistic and assume it
+      // can call anything of the proper type, the same as a CallRef. (We could
+      // look at the possible contents of |target| during the flow, but that
+      // would require special logic like we have for RefCast etc., and the
+      // intrinsics will be lowered away anyhow, so just running after that is
+      // a workaround.)
```

Comment:
> the intrinsics will be lowered away anyhow,

Where are the intrinsics lowered away?

Reply:
The `--intrinsic-lowering` lowering pass does that.